torch/fx/experimental/fuser.py

```diff
@@ -62,3 +70,279 @@ def fuse(model: torch.nn.Module, inplace=False) -> torch.nn.Module:
+                    continue
+            with fx_model.graph.inserting_before(node):
+                mkldnn_args = fx.map_arg(node.args, lambda n: fx_model.graph.call_method('to_mkldnn', (n, )))
+
+            node.args = cast(Tuple[fx.node.Argument], mkldnn_args)
+
+            with fx_model.graph.inserting_after(node):
+                dense_x = fx_model.graph.create_node('call_method', 'to_dense', (node,))
+                node.replace_all_uses_with(dense_x)
+                dense_x.args = (node,)
+
+    # Does pre-conversion of all modules into MKLDNN (when possible)
+    modules_to_mkldnn(list(fx_model.graph.nodes), modules)
+
+    # optimizes all a -> to_dense -> to_mkldnn -> b patterns into a -> b
+    for node in fx_model.graph.nodes:
+        if node.op == 'call_method' and node.target == 'to_dense':
+            prv_node = node.args[0]
+            users = list(node.users)
+            for user in users:
+                if user.op == 'call_method' and user.target == 'to_mkldnn':
+                    user.replace_all_uses_with(prv_node)
+                    fx_model.graph.erase_node(user)
+            if len(node.users) == 0:
+                fx_model.graph.erase_node(node)
+
+    fx_model.recompile()
+
+    num_nodes = len(fx_model.graph.nodes)
+    uf = UnionFind(num_nodes)
+
+    def get_color(n):
+        if hasattr(n, 'color'):
+            return uf.find(n.color)
+        if hasattr(n, 'start_color'):
+            return uf.find(n.start_color)
+        return None
+
+
+    # This code is to find each MKLDNN subgraph. Each MKLDNN subgraph consists
+    # of input nodes (which are only `to_mkldnn` calls), output nodes
+    # (`to_dense` calls), and intermediate nodes, which are run entirely on
+    # MKLDNN layout tensors.
+    #
+    # Specifically, this code does a flood fill on a directed acyclic graph
+    # (DAG), starting from each possible "start node" (i.e: `to_mkldnn` nodes).
+    # If every node only had one input, this would be sufficient. However, in
+    # the case that a node has multiple inputs coming from different start
+    # nodes (i.e. colors), we need to join these 2 colors into 1. That's done
+    # using a Disjoint Set Union.
+    for cur_idx, node in enumerate(fx_model.graph.nodes):
+        if node.op == 'call_method' and node.target == 'to_mkldnn':
+            node.start_color = cur_idx
+            uf.make_set(cur_idx)
+        elif node.op == 'call_method' and node.target == 'to_dense':
+            assert(get_color(node.args[0]) is not None)
+            node.end_color = get_color(node.args[0])
+        else:
+            cur_colors = list(set([get_color(i) for i in node.args if isinstance(i, fx.Node) if get_color(i) is not None]))
+            if len(cur_colors) == 0:
+                continue
+            assert(not any([i is None for i in cur_colors]))
+            node.color = cur_colors[0]
+            for other_color in cur_colors[1:]:
+                uf.join(cur_colors[0], other_color)
```

Comment:
It also seems pretty weird to me that the color that "wins" during a merge is the one that appears first in the argument list. It might make sense to sort `cur_colors` before doing this join s.t. the color that wins is the first in topological order, and thus refers to the index of the earliest `to_mkldnn` node in the subgraph

Reply:
Yeah, I think that's nice.

This stuff doesn't affect any behavior of the pass (the resulting subgraphs are always the same, regardless of what order you choose), but it's nice when debugging the graphs.